### PR TITLE
Pruning of ConvTranspose

### PR DESCRIPTION
**Pull request by AlexanderDokuchaev**

### CODE REVIEW

File: nncf/layer_utils.py [hunk 1]
```diff
@@ -14,6 +14,8 @@ def __getattr__(self, name):
 
 
 class _NNCFModuleMixin:
+    target_compression_weight_dim = 0
```
Comment (vshampor) [inline nncf/layer_utils.py:17]: `target_weight_dim_for_compression`

Comment (vshampor) [reply nncf/layer_utils.py:17]: Also, I expect you to utilize the same mechanism for correct `ConvTranspose` per-channel quantization, perhaps in a separate PR.

**Review by vshampor — CHANGES_REQUESTED** [review]

**Comment by vshampor** [comment]
@mkaglins

File: nncf/layer_utils.py (hunk 1 above)
Comment (AlexanderDokuchaev) [reply nncf/layer_utils.py]: Renamed.
Yes, i will do it in another PR.

File: nncf/pruning/base_algo.py [hunk 2]
```diff
@@ -233,7 +233,7 @@ def pruning_rate_for_filters(minfo: PrunedModuleInfo):
         Calculates sparsity rate for weight filter-wise.
         """
         weight = minfo.module.weight
-        filters_sum = weight.view(weight.size(0), -1).sum(axis=1)
+        filters_sum = weight.view(weight.size(minfo.module.target_weight_dim_for_compression), -1).sum(axis=1)
```
Comment (mkaglins) [inline nncf/pruning/base_algo.py:236]: Here you should transpose weight (in case of target_weight_dim_for_compression != 0) first.
Instead, you will mix filters by reshaping tensor and calculate incorrect filter-wise pruning rate.

File: nncf/pruning/export_helpers.py [hunk 3]
```diff
@@ -169,6 +170,68 @@ def output_prune(self, model: NNCFNetwork, nx_node, graph: NNCFGraph, nx_graph:
...
+        output_mask = nx_node['output_mask']
+        if output_mask is None:
+            return
+
+        bool_mask = torch.tensor(output_mask, dtype=torch.bool)
+        new_num_channels = int(torch.sum(bool_mask))
+
+        nncf_node = graph._nx_node_to_nncf_node(nx_node)
+        node_module = model.get_module_by_scope(nncf_node.op_exec_context.scope_in_model)
+        old_num_clannels = int(node_module.weight.size(0))
```
Comment (mkaglins) [inline nncf/pruning/export_helpers.py:218]: Old num channels are the number of output channels, for ConvTrabspose it should be .size(1)

File: nncf/pruning/filter_pruning/algo.py [hunk 4]
```diff
@@ -128,8 +131,9 @@ def _apply_masks(self):
 
         def _apply_binary_mask_to_module_weight_and_bias(module, mask, module_name=""):
             with torch.no_grad():
+                dim = module.target_weight_dim_for_compression if isinstance(module, _NNCFModuleMixin) else 0
```
Comment (mkaglins) [inline nncf/pruning/filter_pruning/algo.py:134]: Why do you need here this check?
Not all pruned ops are subclasses of `_NNCFModuleMixin`?

Comment (AlexanderDokuchaev) [reply nncf/pruning/filter_pruning/algo.py:134]: Yes, it's need for BatchNorm.

**Comment by AlexanderDokuchaev** [comment]
Jenkins please retry a build

File: nncf/pruning/base_algo.py (hunk 2 above)
Comment (AlexanderDokuchaev) [reply nncf/pruning/base_algo.py]: fixed

File: nncf/layers.py [hunk 5]
```diff
@@ -95,6 +104,9 @@ def from_module(module):
 
 
 class NNCFConvTranspose3d(_NNCFModuleMixin, nn.ConvTranspose3d):
+    op_func_name = "conv_transpose3d"
```
Comment (vshampor) [inline nncf/layers.py:107]: If op_func_name is a required attribute for all NNCF-wrapped modules now, it should be located in a base class. It should also be documented in code there.

**Review by vshampor — CHANGES_REQUESTED** [review]

**Review by mkaglins — APPROVED** [review]

**Review by vshampor — APPROVED** [review]
